relay-server/src/services/project/metrics.rs

```diff
@@ -0,0 +1,333 @@
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.buckets.into_iter()
+    }
+}
+
+impl Buckets {
+    pub fn filter_namespaces(mut self, source: BucketSource) -> Buckets<Filtered> {
+        self.buckets.retain(|bucket| match bucket.name.namespace() {
+            MetricNamespace::Sessions => true,
+            MetricNamespace::Transactions => true,
+            MetricNamespace::Spans => true,
+            MetricNamespace::Profiles => true,
+            MetricNamespace::Custom => true,
+            MetricNamespace::Stats => source == BucketSource::Internal,
+            MetricNamespace::Unsupported => false,
+        });
+
+        self.migrate()
+    }
+}
+
+impl Buckets<Filtered> {
+    pub fn apply_project_state(
+        mut self,
+        outcome_aggregator: &Addr<TrackOutcome>,
+        metric_stats: &MetricStats,
+        state: &ProjectState,
+        scoping: Scoping,
+    ) -> Buckets<WithProjectState> {
+        let mut denied_buckets = Vec::new();
+        let mut disabled_namespace_buckets = Vec::new();
+
+        self.buckets = self
+            .buckets
+            .into_iter()
+            .filter_map(|mut bucket| {
+                if !is_metric_namespace_valid(state, bucket.name.namespace()) {
+                    relay_log::trace!(mri = &*bucket.name, "dropping metric in disabled namespace");
+                    disabled_namespace_buckets.push(bucket);
+                    return None;
+                };
+
+                if let ErrorBoundary::Ok(ref metric_config) = state.config.metrics {
+                    if metric_config.denied_names.is_match(&*bucket.name) {
+                        relay_log::trace!(
+                            mri = &*bucket.name,
+                            "dropping metrics due to block list"
+                        );
+                        denied_buckets.push(bucket);
+                        return None;
+                    }
+
+                    remove_matching_bucket_tags(metric_config, &mut bucket);
+                }
+
+                Some(bucket)
+            })
+            .collect();
+
+        let mode = state.get_extraction_mode();
+        if !disabled_namespace_buckets.is_empty() {
+            utils::reject_metrics(
+                outcome_aggregator,
+                metric_stats,
+                utils::extract_metric_quantities(&disabled_namespace_buckets, mode),
+                scoping,
+                Outcome::Filtered(FilterStatKey::DisabledNamespace),
+                &disabled_namespace_buckets,
+            );
+        }
+
+        if !denied_buckets.is_empty() {
+            utils::reject_metrics(
+                outcome_aggregator,
+                metric_stats,
+                utils::extract_metric_quantities(&denied_buckets, mode),
+                scoping,
+                Outcome::Filtered(FilterStatKey::DeniedName),
+                &denied_buckets,
+            );
+        }
+
+        self.migrate()
```

Comment:
nit: Even though it's not required, I would inlude the type parameter here, just to make extra clear that this is a state transition:

```suggestion
        self.migrate_to::<WithProjectState>()
```

Reply:
I think I'll just remove the method then.